mlir/lib/Dialect/MQTOpt/Transforms/Transpilation/sc/RoutingPass.cpp

```diff
@@ -0,0 +1,1071 @@
+                                   stack.topState().lookupProgram(execIn1),
+                                   stack.topState().lookupHardware(execIn1));
+    });
+
+    stack.topState().remapQubitValue(execIn0, execOut0);
+    stack.topState().remapQubitValue(execIn1, execOut1);
+
+    return WalkResult::advance();
+  }
+
+protected:
+  /**
+   * @brief Insert SWAPs such that @p u is executable.
+   */
+  virtual void makeExecutable(UnitaryInterface op, StateStack& stack,
+                              const Architecture& arch, HardwareIndexPool& pool,
+                              PatternRewriter& rewriter) = 0;
+};
+
+/**
+ * @brief Inserts SWAPs along the shortest path between two hardware
+ * qubits.
+ */
+class NaiveRouter final : public RouterBase {
+protected:
+  void makeExecutable(UnitaryInterface op, StateStack& stack,
+                      const Architecture& arch, HardwareIndexPool& pool,
+                      PatternRewriter& rewriter) final {
+    assert(isTwoQubitGate(op) && "makeExecutable: must be two-qubit gate");
+
+    const auto [qStart, qEnd] = getIns(op);
+    const auto path = getPath(qStart, qEnd, stack.topState(), arch);
+
+    for (std::size_t i = 0; i < path.size() - 2; ++i) {
+      const QubitIndex hardwareIdx0 = path[i];
+      const QubitIndex hardwareIdx1 = path[i + 1];
+
+      const Value qIn0 = stack.topState().lookupHardware(hardwareIdx0);
+      const Value qIn1 = stack.topState().lookupHardware(hardwareIdx1);
+
+      const QubitIndex programIdx0 = stack.topState().lookupProgram(qIn0);
+      const QubitIndex programIdx1 = stack.topState().lookupProgram(qIn1);
+
+      LLVM_DEBUG({
+        llvm::dbgs() << llvm::format(
+            "makeExecutable: swap= s%d/h%d, s%d/h%d <- s%d/h%d, s%d/h%d\n",
+            programIdx1, hardwareIdx0, programIdx0, hardwareIdx1, programIdx0,
+            hardwareIdx0, programIdx1, hardwareIdx1);
+      });
+
+      auto swap = createSwap(op->getLoc(), qIn0, qIn1, rewriter);
+      const auto [qOut0, qOut1] = getOuts(swap);
+
+      rewriter.setInsertionPointAfter(swap);
+      replaceAllUsesInRegionAndChildrenExcept(
+          qIn0, qOut1, swap->getParentRegion(), swap, rewriter);
+      replaceAllUsesInRegionAndChildrenExcept(
+          qIn1, qOut0, swap->getParentRegion(), swap, rewriter);
+
+      stack.recordSwap(programIdx0, programIdx1);
+
+      auto& state = stack.topState();
+      state.swapProgramIndices(qIn0, qIn1);
+      state.remapQubitValue(qIn0, qOut0);
+      state.remapQubitValue(qIn1, qOut1);
+
+      checkZeroUse(qOut0, stack, pool);
+      checkZeroUse(qOut1, stack, pool);
+    }
+  }
+};
+
+/**
+ * @brief The necessary datastructures to route a quantum-classical module.
+ */
+struct RoutingContext {
+  explicit RoutingContext(std::unique_ptr<Architecture> arch,
+                          std::unique_ptr<RouterBase> router,
+                          std::unique_ptr<InitialLayoutGeneratorBase> ilg)
```

Comment:
This is fine for now. I just wanted to note that I am not particularly sure if these will remain as unique pointers here in the near future. Is ownership of these objects really managed by the Context?
For the ILG and the Router, I could imagine that to be true. But for the Device handle, I fail to see that. The resources for that will be managed more centrally (e.g., in the QDMI Driver) and all you'll get here is a handle to be used.
But, as said, nothing to be changed here. So feel free to resolve this after acknowledging the fact.